packages/rln/src/contract/rln_base_contract.ts

```diff
@@ -0,0 +1,707 @@
+import { Logger } from "@waku/utils";
+import { ethers } from "ethers";
+
+import { IdentityCredential } from "../identity.js";
+import { DecryptedCredentials } from "../keystore/types.js";
+
+import { RLN_ABI } from "./abi.js";
+import { DEFAULT_RATE_LIMIT, RATE_LIMIT_PARAMS } from "./constants.js";
+import {
+  CustomQueryOptions,
+  FetchMembersOptions,
+  Member,
+  MembershipInfo,
+  MembershipRegisteredEvent,
+  MembershipState,
+  RLNContractInitOptions
+} from "./types.js";
+
+const log = new Logger("waku:rln:contract:base");
+
+export class RLNBaseContract {
+  public contract: ethers.Contract;
+  private deployBlock: undefined | number;
+  private rateLimit: number;
+
+  protected _members: Map<number, Member> = new Map();
+  private _membersFilter: ethers.EventFilter;
+  private _membershipErasedFilter: ethers.EventFilter;
+  private _membersExpiredFilter: ethers.EventFilter;
+
+  /**
+   * Constructor for RLNBaseContract.
+   * Allows injecting a mocked contract for testing purposes.
+   */
+  public constructor(options: RLNContractInitOptions) {
+    // Initialize members and subscriptions
+    this.fetchMembers()
+      .then(() => {
+        this.subscribeToMembers();
+      })
+      .catch((error) => {
+        log.error("Failed to initialize members", { error });
+      });
+
+    const {
+      address,
+      signer,
+      rateLimit = DEFAULT_RATE_LIMIT,
+      contract
+    } = options;
+
+    this.validateRateLimit(rateLimit);
+
+    this.contract = contract || new ethers.Contract(address, RLN_ABI, signer);
+    this.rateLimit = rateLimit;
+
+    // Initialize event filters
+    this._membersFilter = this.contract.filters.MembershipRegistered();
+    this._membershipErasedFilter = this.contract.filters.MembershipErased();
+    this._membersExpiredFilter = this.contract.filters.MembershipExpired();
+  }
+
+  /**
+   * Gets the current rate limit for this contract instance
+   */
+  public getRateLimit(): number {
```

Comment:
Following same examples below:

```suggestion
  public get rateLimit(): number {
```